packages/react/__tests__/rehydration.js

```diff
@@ -166,3 +198,401 @@ test('initializing another Emotion instance should not move already moved styles
+      </CacheProvider>
+    )
+    const extracted = extractCriticalToChunks(rendered)
+    return {
+      app: extracted.html,
+      styles: constructStyleTagsFromChunks(extracted)
+    }
+  })
+
+  safeQuerySelector('head').innerHTML = styles
+  safeQuerySelector('body').innerHTML = `<div id="root">${app}</div>`
+  expect(safeQuerySelector('html')).toMatchInlineSnapshot(`
+    <html>
+      <head>
+        <style
+          data-emotion="muii-global l6h"
+        >
+          body{color:white;}
+        </style>
+        <style
+          data-emotion="muii 1lrxbo5"
+        >
+          .muii-1lrxbo5{color:hotpink;}
+        </style>
+      </head>
+      <body>
+        <div
+          id="root"
+        >
+          <div
+            class="muii-1lrxbo5"
+          />
+        </div>
+      </body>
+    </html>
+  `)
+
+  resetAllModules()
+  const cache = createCache({ key: 'muii', speedy: true })
+
+  ReactDOM.render(
+    <CacheProvider value={cache}>
+      <Global styles={{ body: { color: 'white' } }} />
+      <Global styles={{ body: { color: 'white' } }} />
+      <div css={{ color: 'hotpink' }} />
+    </CacheProvider>,
+    safeQuerySelector('#root')
+  )
+
+  // it's expected that this contains 2 copies of the same global style
+  // where the second one is added during client hydration
+  // this makes them flushable individually
+  expect(safeQuerySelector('head')).toMatchInlineSnapshot(`
+    <head>
+      <style
+        data-emotion="muii-global"
+        data-s=""
+      >
+        body{color:white;}
+      </style>
+      <style
+        data-emotion="muii-global"
+        data-s=""
+      >
+        
+      </style>
+      <style
+        data-emotion="muii 1lrxbo5"
+        data-s=""
+      >
+        .muii-1lrxbo5{color:hotpink;}
+      </style>
+    </head>
+  `)
+
+  ReactDOM.render(
+    <CacheProvider value={cache}>
+      <Global styles={{ body: { color: 'white' } }} />
+      <div css={{ color: 'hotpink' }} />
+    </CacheProvider>,
+    safeQuerySelector('#root')
+  )
+
+  // this should still have a global style
+  expect(safeQuerySelector('head')).toMatchInlineSnapshot(`
+    <head>
+      <style
+        data-emotion="muii-global"
+        data-s=""
+      >
+        body{color:white;}
+      </style>
+      <style
+        data-emotion="muii 1lrxbo5"
+        data-s=""
+      >
+        .muii-1lrxbo5{color:hotpink;}
+      </style>
+    </head>
+  `)
+
+  ReactDOM.render(
+    <CacheProvider value={cache}>
+      <div css={{ color: 'hotpink' }} />
+    </CacheProvider>,
+    safeQuerySelector('#root')
+  )
+
+  // this should render without a crash
+  expect(safeQuerySelector('head')).toMatchInlineSnapshot(`
+    <head>
+      <style
+        data-emotion="muii 1lrxbo5"
+        data-s=""
+      >
+        .muii-1lrxbo5{color:hotpink;}
+      </style>
+    </head>
+  `)
+})
+
+test('duplicated global styles can be removed safely after rehydrating HTML SSRed with zero config approach', () => {
```

Comment:
I thought that this case could be failing so I've added a test for this but it turned out that it wasn't failing.

I was under the impression that this query would select the same node twice when looking for the same hash:
https://github.com/emotion-js/emotion/blob/9596d32302f389979716852bb63e6e85f11a90c7/packages/react/src/global.js#L98-L100
but:
1. we SSR those style elements as `[0, 1]`
2. they are moved automatically to head when initializing, and the order is preserved: `[0, 1]`
3. when rehydrating the first global it grabs the element 0 and rehydrates it which **moves** that element before all other tags from the main cache so we end up temporarily with `[1 , 0, ...mainCache]` order
4. the second global thus now grabs the element 1 and also moves It before other tags from the main chance and we end up with `[0, 1, ...mainCache]`

I think this works OK and I couldn't find a situation in which it would break but maybe @mitchellhamilton would have some idea for an edge case here that could be tested.